### PR TITLE
Update BindableLayout.cs to fix #5213

### LINKED ISSUE #5213: BindableLayout throws IndexOutOfRange exception during Remove action from ObservableCollection
### Description

I have a StackLayout using the BindableLayout to create a repeater-like control with its ItemSource set to an ObservableCollection. When the items get removed from the collection an IndexOutOfRange exception is thrown. I believe the collection should be sending a reset action instead of remove but as the collection is from realm.io I don't have any control of the action it performs.

### Steps to Reproduce

1. Bind StackLayout BindableLayout to ObservableCollection
2. ObservableCollection triggers CollectionChanged with Action = Remove, OldStartingIndex = 0, and OldItems = All Items
3. IndexOutOfRange exception caused by code at https://github.com/xamarin/Xamarin.Forms/blob/15c6f561e6d51071b7b3cdd9e92d2cbf16edaf2b/Xamarin.Forms.Core/BindableLayout.cs#L228

### Expected Behavior

All children are removed from StackLayout.

### Actual Behavior

IndexOutOfRange exception is thrown

### Basic Information

- Version with issue: 3.5.0.129452
- Last known good version: N/A
- IDE: Visual Studio 2017 Professional
- Platform Target Frameworks: 
  - iOS:  12.1
  - Android: 8.1
- Nuget Packages: Xamarin.Forms
- Affected Devices: All

### PR DESCRIPTION
**Pull request by jsmarcus**
### Description of Change ###

Fixed the ItemSourceCollectionChanged method to not advance the index when removing children during a Remove action.

### Issues Resolved ### 

- fixes #5213 

### API Changes ###
 
 None

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

Tested on Android/iOS.
1. StackLayout with BindableLayout ItemSource set to ObservableCollection.
2. Add 2+ items to the collection.
3. Remove all items from the collection (where the NotifyCollectionChangedAction is Remove, the OldStartingIndex = 0, and the OldItems are all the items in the collection).

### PR Checklist ###

- [X] Has automated tests 
- [X] Rebased on top of the target branch at time of PR
- [X] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs [hunk 1]
```diff
@@ -13,12 +15,12 @@ public class BindableLayoutTests : BaseTestFixture
 		[Test]
 		public void TracksEmpty()
 		{
-			var layout = new StackLayout
+			StackLayout layout = new StackLayout
 			{
```
Comment (andreinitescu) [inline Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs]: Why this change? First, since `new Type` is being used, not using variable type and using var instead is recommended way by Microsoft guidelines. Second, you should not do this kind of changes.

File: Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs [hunk 2]
```diff
@@ -13,12 +15,12 @@ public class BindableLayoutTests : BaseTestFixture
...
 				IsPlatformEnabled = true,
 			};
 
-			var itemsSource = new ObservableCollection<int>();
+			ObservableCollection<int> itemsSource = new ObservableCollection<int>();
```
Comment (andreinitescu) [inline Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs]: Same here

File: Xamarin.Forms.Core/BindableLayout.cs [hunk 3]
```diff
@@ -223,9 +223,9 @@ void ItemsSourceCollectionChanged(object sender, NotifyCollectionChangedEventArg
 					{
 						if (e.OldStartingIndex == -1)
 							goto case NotifyCollectionChangedAction.Reset;
-						for (int i = 0; i < e.OldItems.Count; ++i)
+						for (int i = 0; i < e.OldItems.Count; i++)
```
Comment (andreinitescu) [inline Xamarin.Forms.Core/BindableLayout.cs]: Why change the increment operator order? This is not necessary for the changes you're doing

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/BindableLayout.cs]: not necessary, but we usually `i++` in `for` loops all over the place, so for constancy, I would not reject this change

**Comment by andreinitescu** [comment]
@jsmarcus Thanks for fixing this. If made however a lot of unnecessary changes in the unit tests by changing `var x = new Type` with `Type x = new Type`  which are not necessary.

File: Xamarin.Forms.Core/BindableLayout.cs (hunk 3 above)
Comment (andreinitescu) [reply Xamarin.Forms.Core/BindableLayout.cs]: I see. I prefer `++i` because in some cases I think it can be a tiny bit more performant. For using it in `for` loops like this, it makes no difference.

File: Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs (hunk 1 above)
Comment (jsmarcus) [reply Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs]: Sorry, those changes were unintentional from VS doing a cleanup. I'll fix.

File: Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs (hunk 2 above)
Comment (jsmarcus) [reply Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs]: Sorry, those changes were unintentional from VS doing a cleanup. I'll fix.

Comment (andreinitescu) [reply Xamarin.Forms.Core.UnitTests/BindableLayoutTests.cs]: Oh, I see. No worries, Don't worry about them, it's not too much big of a deal. Thank for your help.

**Comment by jsmarcus** [comment]
Somehow I keep messing up my rebase. I tried a rebase to 3.5.0 since there have been updated to 3.5.0 since my commit but it just seems to have added commits to this PR that are already in 3.5.0.

**Comment by jsmarcus** [comment]
Should I just close this one and try again or is there something I can do to fix it. Sorry for the extra work. Just trying to help.

**Comment by andreinitescu** [comment]
No worries!!! Not sure what are your git issues, but you could try with `reset --soft` and then `push --force`

**Comment by jsmarcus** [comment]
Ok, I got my commits fixed I believe.

**Comment by rmarinho** [comment]
Failing to build, can you check again?

**Comment by jsmarcus** [comment]
Missed some of my changes from the first go around. It's building for me again.

**Comment by jsmarcus** [comment]
It would have been nice to have this in 3.5.0 sr1
